File: agents/rss_fetcher.py
"""
RSS Fetcher Agent
-----------------
Reads active RSS sources (kind='rss') from the DB, fetches each feed with
feedparser, deduplicates via source_url, and inserts new articles as draft posts.

Usage:
    python agents/rss_fetcher.py
    python agents/rss_fetcher.py --tenant=cripto
"""

import argparse
import logging
import time
from datetime import datetime, timezone

import feedparser
from dotenv import load_dotenv

from shared.db import get_db, slugify

load_dotenv()
logging.basicConfig(level=logging.INFO, format="%(levelname)s %(message)s")
log = logging.getLogger(__name__)


def fetch_active_sources(conn, tenant: str | None) -> list[dict]:
    query = "SELECT * FROM sources WHERE kind = 'rss' AND active = 1"
    params: list = []
    if tenant:
        query += " AND tenant = %s"
        params.append(tenant)
    return conn.execute(query, params).fetchall()


def existing_source_urls(conn, tenant: str) -> set[str]:
    rows = conn.execute(
        "SELECT source_url FROM posts WHERE tenant = %s AND source_url IS NOT NULL",
        [tenant],
    ).fetchall()
    return {r["source_url"] for r in rows}


def parse_date(entry) -> datetime | None:
    for attr in ("published_parsed", "updated_parsed"):
        val = getattr(entry, attr, None)
        if val:
            return datetime(*val[:6], tzinfo=timezone.utc)
    return None


def process_feed(conn, source: dict) -> tuple[int, int]:
    log.info("[%s] Fetching %s — %s", source["tenant"], source["label"], source["url"])

    feed = feedparser.parse(source["url"], agent="fluxnews-bot/1.0", request_headers={"Accept": "application/rss+xml"})
    if feed.bozo and not feed.entries:
        log.warning("  ✗ Feed parse error: %s", feed.bozo_exception)
        return 0, 0

    seen = existing_source_urls(conn, source["tenant"])
    inserted = skipped = 0

    for entry in feed.entries:
        url = getattr(entry, "link", None) or getattr(entry, "id", None)
        if not url or url in seen:
            skipped += 1
            continue

        title = getattr(entry, "title", "Sem título").strip()
        excerpt = (getattr(entry, "summary", "") or "")[:300].strip()
        content = (
            getattr(entry, "content", [{}])[0].get("value", "")
            or getattr(entry, "summary", "")
            or excerpt
        )
        author = getattr(entry, "author", "") or ""
        pub_date = parse_date(entry)
        slug = f"{slugify(title)}-{int(time.time() * 1000)}"

        try:
            conn.execute(
                """
                INSERT INTO posts
                    (tenant, title, slug, excerpt, content, category, author,
                     source_url, published_at, status)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, 'draft')
                ON CONFLICT DO NOTHING
                """,
                [
                    source["tenant"], title, slug, excerpt, content,
                    feed.feed.get("title", source["label"]), author,
                    url, pub_date,
                ],
            )
            seen.add(url)
            inserted += 1
        except Exception as exc:
            log.error("  ✗ Insert failed for '%s': %s", title[:50], exc)

    conn.execute(
        "UPDATE sources SET last_fetched_at = NOW() WHERE id = %s",
        [source["id"]],
    )
    conn.commit()
    log.info("  ✓ %d inserted, %d skipped", inserted, skipped)
    return inserted, skipped


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--tenant", default=None)
    args = parser.parse_args()

    log.info("🔄 RSS Fetcher starting…")
    total_in = total_sk = 0

    with get_db() as conn:
        sources = fetch_active_sources(conn, args.tenant)
        if not sources:
            log.info("No active RSS sources found.")
            return
        for source in sources:
            ins, sk = process_feed(conn, source)
            total_in += ins
            total_sk += sk

    log.info("✅ Done. %d inserted, %d skipped.", total_in, total_sk)


if __name__ == "__main__":
    main()

File: packages/agents/src/rss-fetcher.ts
/**
 * RSS Fetcher Agent
 *
 * Reads active RSS sources from the DB, fetches each feed,
 * and inserts new articles as draft posts (skipping duplicates via source_url).
 *
 * Run: pnpm --filter @fluxnews/agents rss
 */

import { db, sources, posts } from '@fluxnews/db';
import { eq, and } from 'drizzle-orm';
import Parser from 'rss-parser';

const parser = new Parser({
  timeout: 10_000,
  headers: { 'User-Agent': 'fluxnews-bot/1.0' },
});

async function fetchRssSources() {
  return db
    .select()
    .from(sources)
    .where(and(eq(sources.kind, 'rss'), eq(sources.active, 1)));
}

async function getExistingSourceUrls(tenant: string): Promise<Set<string>> {
  const existing = await db
    .select({ sourceUrl: posts.sourceUrl })
    .from(posts)
    .where(eq(posts.tenant, tenant));

  return new Set(existing.map((r) => r.sourceUrl).filter(Boolean) as string[]);
}

function slugify(title: string): string {
  return title
    .toLowerCase()
    .normalize('NFD')
    .replace(/[̀-ͯ]/g, '')
    .replace(/[^a-z0-9\s-]/g, '')
    .trim()
    .replace(/\s+/g, '-')
    .slice(0, 80);
}

async function processFeed(source: typeof sources.$inferSelect) {
  console.log(`[${source.tenant}] Fetching ${source.label} — ${source.url}`);

  let feed: Awaited<ReturnType<typeof parser.parseURL>>;
  try {
    feed = await parser.parseURL(source.url);
  } catch (err) {
    console.error(`  ✗ Failed to fetch: ${(err as Error).message}`);
    return { inserted: 0, skipped: 0 };
  }

  const existingUrls = await getExistingSourceUrls(source.tenant);
  let inserted = 0;
  let skipped = 0;

  for (const item of feed.items) {
    const sourceUrl = item.link ?? item.guid;
    if (!sourceUrl || existingUrls.has(sourceUrl)) {
      skipped++;
      continue;
    }

    const title = item.title?.trim() ?? 'Sem título';
    const excerpt = (item.contentSnippet ?? item.summary ?? '').slice(0, 300).trim();
    const slug = `${slugify(title)}-${Date.now()}`;

    try {
      await db.insert(posts).values({
        tenant: source.tenant,
        title,
        slug,
        excerpt,
        content: item.content ?? item['content:encoded'] ?? excerpt,
        category: feed.title ?? source.label,
        author: item.creator ?? item.author ?? '',
        sourceUrl,
        publishedAt: item.pubDate ? new Date(item.pubDate) : null,
        status: 'draft',
      });
      existingUrls.add(sourceUrl);
      inserted++;
    } catch (err) {
      console.error(`  ✗ Insert failed for "${title}": ${(err as Error).message}`);
    }
  }

  // Update last_fetched_at
  await db
    .update(sources)
    .set({ lastFetchedAt: new Date() })
    .where(eq(sources.id, source.id));

  console.log(`  ✓ ${inserted} inserted, ${skipped} skipped`);
  return { inserted, skipped };
}

async function main() {
  console.log('🔄 RSS Fetcher starting…');
  const rssSources = await fetchRssSources();

  if (rssSources.length === 0) {
    console.log('No active RSS sources found. Add sources to the DB first.');
    return;
  }

  let totalInserted = 0;
  let totalSkipped = 0;

  for (const source of rssSources) {
    const { inserted, skipped } = await processFeed(source);
    totalInserted += inserted;
    totalSkipped += skipped;
  }

  console.log(`\n✅ Done. ${totalInserted} articles inserted, ${totalSkipped} skipped.`);
}

main().catch((err) => {
  console.error('Fatal:', err);
  process.exit(1);
});
